### PR TITLE
Cmake super project

### PR DESCRIPTION
**Pull request by ethanrublee**
[CMakeLists.txt](https://github.com/farm-ng/farm-ng-core/files/9643559/CMakeLists.txt)
This cleans up the cmake to be compatible with 3.16 (version on ubuntu 20.04)
Adds some reusable cmake macros

```
# Place this file in a directory with:
# - farm-ng-core # branch cmake-super-project
# - Sophus # branch cmake-super-project
# 
# Then:
#  mkdir build
#  cd build
#  cmake ..
#  make

cmake_minimum_required(VERSION 3.16)
# https://semver.org/
project(farm_ng_all VERSION 0.0.1)


list(APPEND CMAKE_MODULE_PATH 
    ${CMAKE_CURRENT_SOURCE_DIR}/farm-ng-core/cmake)

set(BUILD_SHARED_LIBS ON)
include(farm_ng_default_args)

include(External_fmt)
include(External_expected)
include(External_eigen)
include(External_ceres)
include(External_subdir_farm-ng-core)
include(External_subdir_Sophus)



ExternalProject_Add(farm-ng-core-TestInstallTargets
    DEPENDS farm-ng-core
    PREFIX ${farm_ng_EXT_PREFIX}
    SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/farm-ng-core/cpp/examples/test_install_targets
    BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/farm-ng-core-TestInstallTargets-build
    CMAKE_ARGS ${farm_ng_DEFAULT_ARGS} 
    -DBUILD_FARM_NG_PROTOS=ON
    INSTALL_COMMAND ""
)

ExternalProject_Add(SophusTestInstallTargets
    DEPENDS Sophus
    PREFIX ${farm_ng_EXT_PREFIX}
    SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/Sophus/cpp/examples/test_install_targets
    BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/SophusTestInstallTargets-build
    CMAKE_ARGS ${farm_ng_DEFAULT_ARGS} 
    -DBUILD_FARM_NG_PROTOS=ON
    INSTALL_COMMAND ""
)
```

### CODE REVIEW

File: cpp/farm_ng/core/logging/format.h [hunk 1]
```diff
@@ -23,21 +23,11 @@
 #endif
 
 namespace farm_ng {
-inline std::function<void(std::string const &)> &getLogLineFunction() {
```
Comment (ethanrublee) [inline cpp/farm_ng/core/logging/format.h:26]: @strasdat I moved this to a shared library as this singleton needs to be set across shared library boundaries in normal usecases and the inline static function will not always behave as expected.

Comment (strasdat) [reply cpp/farm_ng/core/logging/format.h:26]: Sounds good. I don't think it is critical to keep farm-ng-core header only.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,16 +1,23 @@
-cmake_minimum_required(VERSION 3.23)
-project(farm_ng_core VERSION 22.99.0)
-
-list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
+cmake_minimum_required(VERSION 3.16)
+# https://semver.org/
+project(farm_ng_core VERSION 0.0.1)
```
Comment (strasdat) [inline CMakeLists.txt:3]: If we follow semver, we should start with version 0.1.0

https://semver.org/#how-should-i-deal-with-revisions-in-the-0yz-initial-development-phase

File: cmake/farm_ng_add_protobufs.cmake [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  farm_ng_add_library(${target}
+    NAMESPACE ${FARM_NG_ADD_PROTOBUFS_NAMESPACE}
+    INCLUDE_DIR ${_proto_output_dir_cpp}
+    HEADERS
+      ${_cpp_out_headers}
+    SOURCES
+      ${_cpp_out_sources})
+
+  target_link_libraries(${target} PUBLIC protobuf::libprotobuf ${FARM_NG_ADD_PROTOBUFS_DEPENDENCIES})
```
Comment (strasdat) [inline cmake/farm_ng_add_protobufs.cmake:50]: This only worked for me on mac (Sophus CI) if 
${FARM_NG_ADD_PROTOBUFS_DEPENDENCIES} = lib::my_lib.

And I believe the following logic won't then be needed nor working correctly:

```
 foreach(_dep_target ${FARM_NG_ADD_PROTOBUFS_DEPENDENCIES})
    list(APPEND DEP_PROTO_INCLUDES  -I ${${_dep_target}_PROTOBUF_IMPORT_DIRS})
  endforeach()
  ```

Comment (strasdat) [reply cmake/farm_ng_add_protobufs.cmake:50]: But I might be wrong. In the end all is good if/when CI passes here: https://github.com/strasdat/Sophus/pull/387

File: cpp/examples/test_install_targets/CMakeLists.txt [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
+cmake_minimum_required(VERSION 3.16)
 project(SophusInstallTargetTests)
 
 set(CMAKE_CXX_STANDARD 17)
 
 include(CMakePackageConfigHelpers)
 include(GNUInstallDirs)
 
-find_package(farm_ng_core 22.99.0 REQUIRED)
+find_package(farm_ng_core 0.0.1 REQUIRED)
```
Comment (strasdat) [inline cpp/examples/test_install_targets/CMakeLists.txt:9]: 0.1.0

File: cmake/farm_ng_add_library.cmake [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    message(FATAL_ERROR "\nPlease specify NAMESPACE in farm_ng_add_library(${target})\n")
+  endif()
+
+  if(NOT DEFINED FARM_NG_ARGS_INCLUDE_DIR)
+    message(FATAL_ERROR "\nPlease specify INCLUDE_DIR in farm_ng_add_library(${target})\n")
+  endif()
+
+
+  if((NOT DEFINED FARM_NG_ARGS_SOURCES) AND (DEFINED FARM_NG_ARGS_HEADERS))
+    set(INTERFACE_OR_PUBLIC INTERFACE)
```
Comment (strasdat) [inline cmake/farm_ng_add_library.cmake:17]: @ethanrublee can we just pass INTERFACE vs. PRIVATE vs. PUBLIC in as an argument? Less magic, and folks google for this and look up what it means.

Comment (strasdat) [reply cmake/farm_ng_add_library.cmake:17]: Or we could have:

farm_ng_add_interface

farm_ng_add_pub_library
etc.

File: CMakeLists.txt (hunk 2 above)
Comment (ethanrublee) [reply CMakeLists.txt]: Thanks fixed.

File: cmake/farm_ng_add_library.cmake (hunk 5 above)
Comment (strasdat) [reply cmake/farm_ng_add_library.cmake:17]: I'm okay to leave this for a follow-up PR though...

File: cmake/farm_ng_add_protobufs.cmake [hunk 6]
```diff
@@ -0,0 +1,61 @@
+macro(farm_ng_add_protobufs target)
+  set(multi_value_args NAMESPACE PROTO_FILES DEPENDENCIES)
+  cmake_parse_arguments(FARM_NG_ADD_PROTOBUFS "" "" "${multi_value_args}" ${ARGN})
+  set(${target}_PROTOBUF_IMPORT_DIRS ${CMAKE_CURRENT_SOURCE_DIR} CACHE STRING "Path to this project's protobuf sources")
+  # foreach(_dep_target ${FARM_NG_ADD_PROTOBUFS_DEPENDENCIES})
```
Comment (strasdat) [inline cmake/farm_ng_add_protobufs.cmake:5]: remove

File: cmake/farm_ng_coreConfig.cmake.in [hunk 7]
```diff
@@ -4,3 +4,9 @@ include(CMakeFindDependencyMacro)
 find_dependency(fmt REQUIRED)
 
 include ("${CMAKE_CURRENT_LIST_DIR}/farm_ng_coreTargets.cmake")
+
+#list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
```
Comment (strasdat) [inline cmake/farm_ng_coreConfig.cmake.in:8]: remove?

File: python/.gitignore [hunk 8]
```diff
@@ -0,0 +1,5 @@
+*venv*
```
Comment (strasdat) [inline python/.gitignore:1]: we have a "python" and a "py" folder. Should only use one or the other.

**Pull request merged**